include/linux/kfence.h

```diff
@@ -84,6 +84,21 @@ kfence_sampled_alloc_with_size(struct kmem_cache *s, gfp_t flags, size_t size)
 	return kfence_alloc_with_size(s, size, flags);
 }
 
+#elif defined(CONFIG_KFENCE_STATIC_BRANCH)
+
+#include <linux/static_key.h>
+
+extern struct static_key_false kfence_allocation_key;
+
+void *kfence_alloc_with_size(struct kmem_cache *s, size_t size, gfp_t flags);
+static __always_inline void *
+kfence_sampled_alloc_with_size(struct kmem_cache *s, gfp_t flags, size_t size)
+{
+	return static_branch_unlikely(&kfence_allocation_key) ?
```

Comment:
Does it also eliminate the branch on kfence_sampled_alloc_with_size return value in slab_alloc_node?

Reply:
Please post disasm of kmalloc fast path.

Reply:
Here is __kmalloc_node. You can see that kfence_alloc_with_size is simply jumped over by default.

```
ffffffff811e2760 <__kmalloc_node>:
ffffffff811e2760:       41 57                   push   %r15
ffffffff811e2762:       41 56                   push   %r14
ffffffff811e2764:       41 89 d6                mov    %edx,%r14d
ffffffff811e2767:       41 55                   push   %r13
ffffffff811e2769:       49 89 fd                mov    %rdi,%r13
ffffffff811e276c:       41 54                   push   %r12
ffffffff811e276e:       55                      push   %rbp
ffffffff811e276f:       89 f5                   mov    %esi,%ebp
ffffffff811e2771:       53                      push   %rbx
ffffffff811e2772:       48 83 ec 08             sub    $0x8,%rsp
ffffffff811e2776:       48 81 ff 00 20 00 00    cmp    $0x2000,%rdi
ffffffff811e277d:       0f 87 17 02 00 00       ja     ffffffff811e299a <__kmalloc_node+0x23a>
ffffffff811e2783:       e8 88 ad fb ff          callq  ffffffff8119d510 <kmalloc_slab>
ffffffff811e2788:       49 89 c7                mov    %rax,%r15
ffffffff811e278b:       48 83 f8 10             cmp    $0x10,%rax
ffffffff811e278f:       0f 86 92 02 00 00       jbe    ffffffff811e2a27 <__kmalloc_node+0x2c7>
ffffffff811e2795:       48 8b 44 24 38          mov    0x38(%rsp),%rax
ffffffff811e279a:       48 89 04 24             mov    %rax,(%rsp)
ffffffff811e279e:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
ffffffff811e27a3:       8b 35 27 c7 39 01       mov    0x139c727(%rip),%esi        # ffffffff8257eed0 <gfp_allowed_mask>
ffffffff811e27a9:       4c 89 ff                mov    %r15,%rdi
ffffffff811e27ac:       21 ee                   and    %ebp,%esi
ffffffff811e27ae:       e8 6d 8b fb ff          callq  ffffffff8119b320 <should_failslab>
ffffffff811e27b3:       85 c0                   test   %eax,%eax
ffffffff811e27b5:       0f 85 d7 01 00 00       jne    ffffffff811e2992 <__kmalloc_node+0x232>
ffffffff811e27bb:       4d 85 ff                test   %r15,%r15
ffffffff811e27be:       0f 84 ce 01 00 00       je     ffffffff811e2992 <__kmalloc_node+0x232>
ffffffff811e27c4:       49 8b 0f                mov    (%r15),%rcx
ffffffff811e27c7:       65 48 8b 51 08          mov    %gs:0x8(%rcx),%rdx
ffffffff811e27cc:       48 89 c8                mov    %rcx,%rax
ffffffff811e27cf:       65 48 03 05 91 eb e2    add    %gs:0x7ee2eb91(%rip),%rax        # 11368 <this_cpu_off>
ffffffff811e27d6:       7e
ffffffff811e27d7:       48 8b 70 08             mov    0x8(%rax),%rsi
ffffffff811e27db:       48 39 f2                cmp    %rsi,%rdx
ffffffff811e27de:       75 e7                   jne    ffffffff811e27c7 <__kmalloc_node+0x67>
ffffffff811e27e0:       4c 8b 20                mov    (%rax),%r12
ffffffff811e27e3:       4d 85 e4                test   %r12,%r12
ffffffff811e27e6:       74 1e                   je     ffffffff811e2806 <__kmalloc_node+0xa6>
ffffffff811e27e8:       41 83 fe ff             cmp    $0xffffffff,%r14d
ffffffff811e27ec:       0f 84 14 01 00 00       je     ffffffff811e2906 <__kmalloc_node+0x1a6>
ffffffff811e27f2:       48 8b 40 10             mov    0x10(%rax),%rax
ffffffff811e27f6:       48 8b 00                mov    (%rax),%rax
ffffffff811e27f9:       48 c1 e8 3a             shr    $0x3a,%rax
ffffffff811e27fd:       41 39 c6                cmp    %eax,%r14d
ffffffff811e2800:       0f 84 00 01 00 00       je     ffffffff811e2906 <__kmalloc_node+0x1a6>
ffffffff811e2806:       9c                      pushfq
ffffffff811e2807:       5b                      pop    %rbx
ffffffff811e2808:       fa                      cli
ffffffff811e2809:       4d 8b 27                mov    (%r15),%r12
ffffffff811e280c:       e8 ef 17 9b 00          callq  ffffffff81b94000 <debug_smp_processor_id>
ffffffff811e2811:       48 8b 0c 24             mov    (%rsp),%rcx
ffffffff811e2815:       44 89 f2                mov    %r14d,%edx
ffffffff811e2818:       89 ee                   mov    %ebp,%esi
ffffffff811e281a:       89 c0                   mov    %eax,%eax
ffffffff811e281c:       4c 89 ff                mov    %r15,%rdi
ffffffff811e281f:       4c 8b 04 c5 e0 f6 2d    mov    -0x7dd20920(,%rax,8),%r8
ffffffff811e2826:       82
ffffffff811e2827:       4d 01 e0                add    %r12,%r8
ffffffff811e282a:       e8 b1 f0 ff ff          callq  ffffffff811e18e0 <___slab_alloc.constprop.0>
ffffffff811e282f:       49 89 c4                mov    %rax,%r12
ffffffff811e2832:       53                      push   %rbx
ffffffff811e2833:       9d                      popfq
ffffffff811e2834:       48 89 c7                mov    %rax,%rdi
ffffffff811e2837:       e8 04 3e 00 00          callq  ffffffff811e6640 <is_kfence_addr>
ffffffff811e283c:       84 c0                   test   %al,%al
ffffffff811e283e:       75 40                   jne    ffffffff811e2880 <__kmalloc_node+0x120>
ffffffff811e2840:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
ffffffff811e2845:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
ffffffff811e284a:       f7 c5 00 01 00 00       test   $0x100,%ebp
ffffffff811e2850:       74 2e                   je     ffffffff811e2880 <__kmalloc_node+0x120>
ffffffff811e2852:       4d 85 e4                test   %r12,%r12
ffffffff811e2855:       74 29                   je     ffffffff811e2880 <__kmalloc_node+0x120>
ffffffff811e2857:       41 8b 57 1c             mov    0x1c(%r15),%edx
ffffffff811e285b:       31 f6                   xor    %esi,%esi
ffffffff811e285d:       4c 89 e7                mov    %r12,%rdi
ffffffff811e2860:       e8 1b 2a 26 00          callq  ffffffff81445280 <__memset>
ffffffff811e2865:       eb 19                   jmp    ffffffff811e2880 <__kmalloc_node+0x120>
ffffffff811e2867:       89 ea                   mov    %ebp,%edx
ffffffff811e2869:       4c 89 ee                mov    %r13,%rsi
ffffffff811e286c:       4c 89 ff                mov    %r15,%rdi
ffffffff811e286f:       e8 0c 4f 00 00          callq  ffffffff811e7780 <kfence_alloc_with_size>
ffffffff811e2874:       49 89 c4                mov    %rax,%r12
ffffffff811e2877:       48 85 c0                test   %rax,%rax
ffffffff811e287a:       0f 84 23 ff ff ff       je     ffffffff811e27a3 <__kmalloc_node+0x43>
ffffffff811e2880:       45 8b 7f 18             mov    0x18(%r15),%r15d
ffffffff811e2884:       48 8b 74 24 38          mov    0x38(%rsp),%rsi
ffffffff811e2889:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
ffffffff811e288e:       48 83 c4 08             add    $0x8,%rsp
ffffffff811e2892:       4c 89 e0                mov    %r12,%rax
ffffffff811e2895:       5b                      pop    %rbx
ffffffff811e2896:       5d                      pop    %rbp
ffffffff811e2897:       41 5c                   pop    %r12
ffffffff811e2899:       41 5d                   pop    %r13
ffffffff811e289b:       41 5e                   pop    %r14
ffffffff811e289d:       41 5f                   pop    %r15
ffffffff811e289f:       c3                      retq
ffffffff811e28a0:       65 8b 05 b9 ea e2 7e    mov    %gs:0x7ee2eab9(%rip),%eax        # 11360 <cpu_number>
ffffffff811e28a7:       89 c0                   mov    %eax,%eax
ffffffff811e28a9:       48 0f a3 05 57 be 39    bt     %rax,0x139be57(%rip)        # ffffffff8257e708 <__cpu_online_mask>
ffffffff811e28b0:       01
ffffffff811e28b1:       73 db                   jae    ffffffff811e288e <__kmalloc_node+0x12e>
ffffffff811e28b3:       65 ff 05 06 44 e3 7e    incl   %gs:0x7ee34406(%rip)        # 16cc0 <__preempt_count>
ffffffff811e28ba:       48 8b 1d 87 84 38 01    mov    0x1388487(%rip),%rbx        # ffffffff8256ad48 <__tracepoint_kmalloc_node+0x28>
ffffffff811e28c1:       48 85 db                test   %rbx,%rbx
ffffffff811e28c4:       74 30                   je     ffffffff811e28f6 <__kmalloc_node+0x196>
ffffffff811e28c6:       48 8b 03                mov    (%rbx),%rax
ffffffff811e28c9:       48 8b 7b 08             mov    0x8(%rbx),%rdi
ffffffff811e28cd:       41 56                   push   %r14
ffffffff811e28cf:       48 83 c3 18             add    $0x18,%rbx
ffffffff811e28d3:       4c 89 e2                mov    %r12,%rdx
ffffffff811e28d6:       41 89 e9                mov    %ebp,%r9d
ffffffff811e28d9:       4d 89 f8                mov    %r15,%r8
ffffffff811e28dc:       4c 89 e9                mov    %r13,%rcx
ffffffff811e28df:       48 89 74 24 08          mov    %rsi,0x8(%rsp)
ffffffff811e28e4:       e8 17 e6 c1 00          callq  ffffffff81e00f00 <__x86_indirect_thunk_rax>
ffffffff811e28e9:       48 8b 03                mov    (%rbx),%rax
ffffffff811e28ec:       5a                      pop    %rdx
ffffffff811e28ed:       48 8b 34 24             mov    (%rsp),%rsi
ffffffff811e28f1:       48 85 c0                test   %rax,%rax
ffffffff811e28f4:       75 d3                   jne    ffffffff811e28c9 <__kmalloc_node+0x169>
ffffffff811e28f6:       65 ff 0d c3 43 e3 7e    decl   %gs:0x7ee343c3(%rip)        # 16cc0 <__preempt_count>
ffffffff811e28fd:       75 8f                   jne    ffffffff811e288e <__kmalloc_node+0x12e>
ffffffff811e28ff:       e8 cb f5 e1 ff          callq  ffffffff81001ecf <preempt_schedule_notrace_thunk>
ffffffff811e2904:       eb 88                   jmp    ffffffff811e288e <__kmalloc_node+0x12e>
ffffffff811e2906:       41 8b 47 20             mov    0x20(%r15),%eax
ffffffff811e290a:       49 8b 3f                mov    (%r15),%rdi
ffffffff811e290d:       48 8d 4a 40             lea    0x40(%rdx),%rcx
ffffffff811e2911:       49 8b 1c 04             mov    (%r12,%rax,1),%rbx
ffffffff811e2915:       4c 89 e0                mov    %r12,%rax
ffffffff811e2918:       48 8d 37                lea    (%rdi),%rsi
ffffffff811e291b:       e8 a0 fe 25 00          callq  ffffffff814427c0 <this_cpu_cmpxchg16b_emu>
ffffffff811e2920:       84 c0                   test   %al,%al
ffffffff811e2922:       0f 84 9c fe ff ff       je     ffffffff811e27c4 <__kmalloc_node+0x64>
ffffffff811e2928:       41 8b 47 20             mov    0x20(%r15),%eax
ffffffff811e292c:       0f 18 0c 03             prefetcht0 (%rbx,%rax,1)
ffffffff811e2930:       e9 0b ff ff ff          jmpq   ffffffff811e2840 <__kmalloc_node+0xe0>
ffffffff811e2935:       49 83 7f 40 00          cmpq   $0x0,0x40(%r15)
ffffffff811e293a:       0f 85 40 ff ff ff       jne    ffffffff811e2880 <__kmalloc_node+0x120>
ffffffff811e2940:       41 f7 47 08 00 08 08    testl  $0x80800,0x8(%r15)
ffffffff811e2947:       00
ffffffff811e2948:       0f 84 04 ff ff ff       je     ffffffff811e2852 <__kmalloc_node+0xf2>
ffffffff811e294e:       f7 c5 00 01 00 00       test   $0x100,%ebp
ffffffff811e2954:       0f 84 26 ff ff ff       je     ffffffff811e2880 <__kmalloc_node+0x120>
ffffffff811e295a:       e9 f3 fe ff ff          jmpq   ffffffff811e2852 <__kmalloc_node+0xf2>
ffffffff811e295f:       49 83 7f 40 00          cmpq   $0x0,0x40(%r15)
ffffffff811e2964:       0f 85 db fe ff ff       jne    ffffffff811e2845 <__kmalloc_node+0xe5>
ffffffff811e296a:       41 f7 47 08 00 08 08    testl  $0x80800,0x8(%r15)
ffffffff811e2971:       00
ffffffff811e2972:       0f 85 cd fe ff ff       jne    ffffffff811e2845 <__kmalloc_node+0xe5>
ffffffff811e2978:       4d 85 e4                test   %r12,%r12
ffffffff811e297b:       0f 84 c4 fe ff ff       je     ffffffff811e2845 <__kmalloc_node+0xe5>
ffffffff811e2981:       41 8b 47 20             mov    0x20(%r15),%eax
ffffffff811e2985:       49 c7 04 04 00 00 00    movq   $0x0,(%r12,%rax,1)
ffffffff811e298c:       00
ffffffff811e298d:       e9 b3 fe ff ff          jmpq   ffffffff811e2845 <__kmalloc_node+0xe5>
ffffffff811e2992:       45 31 e4                xor    %r12d,%r12d
ffffffff811e2995:       e9 e6 fe ff ff          jmpq   ffffffff811e2880 <__kmalloc_node+0x120>
ffffffff811e299a:       e8 31 bf ff ff          callq  ffffffff811de8d0 <kmalloc_large_node>
ffffffff811e299f:       4c 8b 7c 24 38          mov    0x38(%rsp),%r15
ffffffff811e29a4:       49 89 c4                mov    %rax,%r12
ffffffff811e29a7:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
ffffffff811e29ac:       e9 dd fe ff ff          jmpq   ffffffff811e288e <__kmalloc_node+0x12e>
ffffffff811e29b1:       65 8b 05 a8 e9 e2 7e    mov    %gs:0x7ee2e9a8(%rip),%eax        # 11360 <cpu_number>
ffffffff811e29b8:       89 c0                   mov    %eax,%eax
ffffffff811e29ba:       48 0f a3 05 46 bd 39    bt     %rax,0x139bd46(%rip)        # ffffffff8257e708 <__cpu_online_mask>
ffffffff811e29c1:       01
ffffffff811e29c2:       0f 83 c6 fe ff ff       jae    ffffffff811e288e <__kmalloc_node+0x12e>
ffffffff811e29c8:       65 ff 05 f1 42 e3 7e    incl   %gs:0x7ee342f1(%rip)        # 16cc0 <__preempt_count>
ffffffff811e29cf:       48 8b 1d 72 83 38 01    mov    0x1388372(%rip),%rbx        # ffffffff8256ad48 <__tracepoint_kmalloc_node+0x28>
ffffffff811e29d6:       48 85 db                test   %rbx,%rbx
ffffffff811e29d9:       0f 84 17 ff ff ff       je     ffffffff811e28f6 <__kmalloc_node+0x196>
ffffffff811e29df:       4c 89 ef                mov    %r13,%rdi
ffffffff811e29e2:       41 b8 00 10 00 00       mov    $0x1000,%r8d
ffffffff811e29e8:       e8 43 be ff ff          callq  ffffffff811de830 <get_order>
ffffffff811e29ed:       89 c1                   mov    %eax,%ecx
ffffffff811e29ef:       48 8b 03                mov    (%rbx),%rax
ffffffff811e29f2:       49 d3 e0                shl    %cl,%r8
ffffffff811e29f5:       48 8b 7b 08             mov    0x8(%rbx),%rdi
ffffffff811e29f9:       41 56                   push   %r14
ffffffff811e29fb:       48 83 c3 18             add    $0x18,%rbx
ffffffff811e29ff:       4c 89 e9                mov    %r13,%rcx
ffffffff811e2a02:       41 89 e9                mov    %ebp,%r9d
ffffffff811e2a05:       4c 89 e2                mov    %r12,%rdx
ffffffff811e2a08:       4c 89 fe                mov    %r15,%rsi
ffffffff811e2a0b:       4c 89 44 24 08          mov    %r8,0x8(%rsp)
ffffffff811e2a10:       e8 eb e4 c1 00          callq  ffffffff81e00f00 <__x86_indirect_thunk_rax>
ffffffff811e2a15:       48 8b 03                mov    (%rbx),%rax
ffffffff811e2a18:       59                      pop    %rcx
ffffffff811e2a19:       4c 8b 04 24             mov    (%rsp),%r8
ffffffff811e2a1d:       48 85 c0                test   %rax,%rax
ffffffff811e2a20:       75 d3                   jne    ffffffff811e29f5 <__kmalloc_node+0x295>
ffffffff811e2a22:       e9 cf fe ff ff          jmpq   ffffffff811e28f6 <__kmalloc_node+0x196>
ffffffff811e2a27:       49 89 c4                mov    %rax,%r12
ffffffff811e2a2a:       e9 5f fe ff ff          jmpq   ffffffff811e288e <__kmalloc_node+0x12e>
ffffffff811e2a2f:       90                      nop


```

Reply:
There should be even no jump I think.
The static key should result in a 5-byte nop with is later overwritten with some jump.

I did:
```
+DEFINE_STATIC_KEY_FALSE(kfence_allocation_key);
+
 ssize_t slabinfo_write(struct file *file, const char __user *buffer,
                       size_t count, loff_t *ppos)
 {
+       if (static_branch_unlikely(&kfence_allocation_key))
+               kfree(0);
        return -EIO;
 }
```

and got:
```
ffffffff811d0220 <slabinfo_write>:
ffffffff811d0220:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
ffffffff811d0225:       48 c7 c0 fb ff ff ff    mov    $0xfffffffffffffffb,%rax
ffffffff811d022c:       c3                      retq   
ffffffff811d022d:       31 ff                   xor    %edi,%edi
ffffffff811d022f:       e8 ec ed ff ff          callq  ffffffff811cf020 <kfree>
ffffffff811d0234:       48 c7 c0 fb ff ff ff    mov    $0xfffffffffffffffb,%rax
ffffffff811d023b:       c3                      retq   
```

The static key branch overhead on fast path is only 1 nop instruction.

Can we get the same? Or do we have it already?
Maybe we need more unlikely annotations?

Reply:
Ah I looked at the wrong jmp. I think we already have that, see the nop at ffffffff811e279e.